File: engines/python/agentce/assertions.py
"""Assertions: one machine-readable record per (control, subject) (SPEC §9.4, §9.2).

``assertions.json`` is the source of truth from which report.md/html, OSCAL, and SARIF are rendered; a
translation never changes it, and after RFC 8785 canonicalisation it is byte-identical across engines.
Every non-``not_applicable`` verdict that asserts support (``conformant``, ``non-conformant``,
``partial``) must cite at least one evidence pointer; the report generator refuses to emit one that
does not (DC-5).
"""

from __future__ import annotations

from dataclasses import dataclass, field
from typing import Any

from .errors import AgentceError
from .exit_codes import ExitCode

OUTCOMES = (
    "conformant",
    "non-conformant",
    "partial",
    "not_applicable",
    "not_assessed",
    "insufficient_evidence",
)

#: Outcomes that assert support and therefore require evidence pointers (DC-5).
_REQUIRE_EVIDENCE = frozenset({"conformant", "non-conformant", "partial"})


@dataclass(frozen=True)
class EvidencePointer:
    ref: str
    digest: str
    source_class: str

    def to_json(self) -> dict[str, str]:
        return {
            "ref": self.ref,
            "digest": self.digest,
            "source_class": self.source_class,
        }

    @classmethod
    def from_json(cls, data: dict[str, Any]) -> EvidencePointer:
        return cls(
            ref=str(data.get("ref", "")),
            digest=str(data.get("digest", "")),
            source_class=str(data.get("source_class", "")),
        )


@dataclass
class Assertion:
    control: str
    control_version: str
    subject: str
    outcome: str
    rung: int
    mode: str
    window: tuple[str, str]
    population: tuple[int, int]
    expectations: list[dict[str, str]] = field(default_factory=list)
    violations: list[dict[str, str]] = field(default_factory=list)
    evidence: list[EvidencePointer] = field(default_factory=list)
    source_class_satisfied: bool | None = None
    evidence_strength: str | None = None
    deviation: str | None = None
    crosswalk: list[dict[str, Any]] = field(default_factory=list)

    def to_json(self) -> dict[str, Any]:
        record: dict[str, Any] = {
            "control": self.control,
            "control_version": self.control_version,
            "subject": self.subject,
            "outcome": self.outcome,
            "rung": self.rung,
            "mode": self.mode,
            "window": {"start": self.window[0], "end": self.window[1]},
            "population": {
                "applicable": self.population[0],
                "failed": self.population[1],
            },
        }
        if self.expectations:
            record["expectations"] = self.expectations
        if self.violations:
            record["violations"] = self.violations
        if self.evidence:
            record["evidence"] = [e.to_json() for e in self.evidence]
        if self.source_class_satisfied is not None:
            record["source_class_satisfied"] = self.source_class_satisfied
        if self.evidence_strength is not None:
            record["evidence_strength"] = self.evidence_strength
        if self.deviation is not None:
            record["deviation"] = self.deviation
        if self.crosswalk:
            record["crosswalk"] = self.crosswalk
        return record

    @classmethod
    def from_json(cls, data: dict[str, Any]) -> Assertion:
        """Reconstruct an assertion from its JSON form (the inverse of ``to_json``), for re-rendering
        a report from a committed ``assertions.json`` (SPEC §9.4)."""
        window = data.get("window", {})
        population = data.get("population", {})
        return cls(
            control=str(data.get("control", "")),
            control_version=str(data.get("control_version", "")),
            subject=str(data.get("subject", "")),
            outcome=str(data.get("outcome", "")),
            rung=int(data.get("rung", 0)),
            mode=str(data.get("mode", "")),
            window=(str(window.get("start", "")), str(window.get("end", ""))),
            population=(
                int(population.get("applicable", 0)),
                int(population.get("failed", 0)),
            ),
            expectations=list(data.get("expectations", [])),
            violations=list(data.get("violations", [])),
            evidence=[EvidencePointer.from_json(e) for e in data.get("evidence", [])],
            source_class_satisfied=data.get("source_class_satisfied"),
            evidence_strength=data.get("evidence_strength"),
            deviation=data.get("deviation"),
            crosswalk=list(data.get("crosswalk", [])),
        )


def aggregate(assertions: list[Assertion]) -> dict[str, int]:
    """Return the six-outcome counts (SPEC §9.2); every outcome is present, even at zero."""
    counts = dict.fromkeys(OUTCOMES, 0)
    for assertion in assertions:
        counts[assertion.outcome] = counts.get(assertion.outcome, 0) + 1
    return counts


def check_dc5(assertions: list[Assertion]) -> None:
    """Refuse to emit a supporting verdict with no evidence pointer (DC-5); raise on the first."""
    for assertion in assertions:
        if assertion.outcome in _REQUIRE_EVIDENCE and not assertion.evidence:
            raise AgentceError(
                key="report.missing_evidence_pointer",
                cause=(
                    f"control {assertion.control} on {assertion.subject} is {assertion.outcome} "
                    "but cites no evidence pointer."
                ),
                fix="every conformant, non-conformant, or partial outcome must cite evidence (DC-5).",
                exit_code=int(ExitCode.INPUT_ERROR),
            )

File: engines/typescript/src/assertions.ts
/**
 * Assertions: one machine-readable record per (control, subject) (SPEC §9.4, §9.2).
 *
 * `assertions.json` is the source of truth from which report.md/html, OSCAL, and SARIF are rendered; a
 * translation never changes it, and after RFC 8785 canonicalisation it is byte-identical across
 * engines. Every non-`not_applicable` verdict that asserts support (`conformant`, `non-conformant`,
 * `partial`) must cite at least one evidence pointer; the report generator refuses to emit one that
 * does not (DC-5). This is a faithful port of the Python reference.
 */

import { AgentceError } from "./errors";
import { ExitCode } from "./exitCodes";

export const OUTCOMES = [
  "conformant",
  "non-conformant",
  "partial",
  "not_applicable",
  "not_assessed",
  "insufficient_evidence",
] as const;

/** Outcomes that assert support and therefore require evidence pointers (DC-5). */
const REQUIRE_EVIDENCE = new Set(["conformant", "non-conformant", "partial"]);

export interface EvidencePointer {
  ref: string;
  digest: string;
  sourceClass: string;
}

export function evidencePointerToJson(pointer: EvidencePointer): Record<string, string> {
  return { ref: pointer.ref, digest: pointer.digest, source_class: pointer.sourceClass };
}

export interface Assertion {
  control: string;
  controlVersion: string;
  subject: string;
  outcome: string;
  rung: number;
  mode: string;
  window: [string, string];
  population: [number, number];
  severity: string;
  family: string;
  expectations: Array<Record<string, string>>;
  violations: Array<Record<string, string>>;
  evidence: EvidencePointer[];
  sourceClassSatisfied: boolean | null;
  evidenceStrength: string | null;
  deviation: string | null;
  crosswalk: Array<Record<string, string | boolean>>;
}

/** Build an assertion, defaulting the optional collections and nullable fields (mirrors the dataclass). */
export function makeAssertion(
  fields: Pick<
    Assertion,
    | "control"
    | "controlVersion"
    | "subject"
    | "outcome"
    | "rung"
    | "mode"
    | "window"
    | "population"
    | "severity"
    | "family"
  > &
    Partial<Assertion>,
): Assertion {
  return {
    expectations: [],
    violations: [],
    evidence: [],
    sourceClassSatisfied: null,
    evidenceStrength: null,
    deviation: null,
    crosswalk: [],
    ...fields,
  };
}

export function assertionToJson(assertion: Assertion): Record<string, unknown> {
  const record: Record<string, unknown> = {
    control: assertion.control,
    control_version: assertion.controlVersion,
    subject: assertion.subject,
    outcome: assertion.outcome,
    rung: assertion.rung,
    mode: assertion.mode,
    window: { start: assertion.window[0], end: assertion.window[1] },
    population: { applicable: assertion.population[0], failed: assertion.population[1] },
    severity: assertion.severity,
    family: assertion.family,
  };
  if (assertion.expectations.length > 0) {
    record.expectations = assertion.expectations;
  }
  if (assertion.violations.length > 0) {
    record.violations = assertion.violations;
  }
  if (assertion.evidence.length > 0) {
    record.evidence = assertion.evidence.map(evidencePointerToJson);
  }
  if (assertion.sourceClassSatisfied !== null) {
    record.source_class_satisfied = assertion.sourceClassSatisfied;
  }
  if (assertion.evidenceStrength !== null) {
    record.evidence_strength = assertion.evidenceStrength;
  }
  if (assertion.deviation !== null) {
    record.deviation = assertion.deviation;
  }
  if (assertion.crosswalk.length > 0) {
    record.crosswalk = assertion.crosswalk;
  }
  return record;
}

function isRecord(value: unknown): value is Record<string, unknown> {
  return typeof value === "object" && value !== null && !Array.isArray(value);
}

function evidencePointerFromJson(data: unknown): EvidencePointer {
  const d = isRecord(data) ? data : {};
  return {
    ref: String(d.ref ?? ""),
    digest: String(d.digest ?? ""),
    sourceClass: String(d.source_class ?? ""),
  };
}

/** Reconstruct an assertion from its JSON form (the inverse of {@link assertionToJson}), for
 * re-rendering a report from a committed `assertions.json` (SPEC §9.4). */
export function assertionFromJson(data: unknown): Assertion {
  const d = isRecord(data) ? data : {};
  const window = isRecord(d.window) ? d.window : {};
  const population = isRecord(d.population) ? d.population : {};
  return {
    control: String(d.control ?? ""),
    controlVersion: String(d.control_version ?? ""),
    subject: String(d.subject ?? ""),
    outcome: String(d.outcome ?? ""),
    rung: Number(d.rung ?? 0),
    mode: String(d.mode ?? ""),
    window: [String(window.start ?? ""), String(window.end ?? "")],
    population: [Number(population.applicable ?? 0), Number(population.failed ?? 0)],
    severity: String(d.severity ?? ""),
    family: String(d.family ?? ""),
    expectations: Array.isArray(d.expectations)
      ? (d.expectations as Array<Record<string, string>>)
      : [],
    violations: Array.isArray(d.violations) ? (d.violations as Array<Record<string, string>>) : [],
    evidence: Array.isArray(d.evidence) ? d.evidence.map(evidencePointerFromJson) : [],
    sourceClassSatisfied:
      typeof d.source_class_satisfied === "boolean" ? d.source_class_satisfied : null,
    evidenceStrength: typeof d.evidence_strength === "string" ? d.evidence_strength : null,
    deviation: typeof d.deviation === "string" ? d.deviation : null,
    crosswalk: Array.isArray(d.crosswalk)
      ? (d.crosswalk as Array<Record<string, string | boolean>>)
      : [],
  };
}

/** Return the six-outcome counts (SPEC §9.2); every outcome is present, even at zero. */
export function aggregate(assertions: Assertion[]): Record<string, number> {
  const counts: Record<string, number> = {};
  for (const outcome of OUTCOMES) {
    counts[outcome] = 0;
  }
  for (const assertion of assertions) {
    counts[assertion.outcome] = (counts[assertion.outcome] ?? 0) + 1;
  }
  return counts;
}

/** Refuse to emit a supporting verdict with no evidence pointer (DC-5); raise on the first. */
export function checkDc5(assertions: Assertion[]): void {
  for (const assertion of assertions) {
    if (REQUIRE_EVIDENCE.has(assertion.outcome) && assertion.evidence.length === 0) {
      throw new AgentceError(
        "report.missing_evidence_pointer",
        `control ${assertion.control} on ${assertion.subject} is ${assertion.outcome} but cites no evidence pointer.`,
        "every conformant, non-conformant, or partial outcome must cite evidence (DC-5).",
        ExitCode.INPUT_ERROR,
      );
    }
  }
}
